File: leetcode_c++/src/2069.cpp
#include <algorithm>
#include <array>
#include <iostream>
#include <iterator>
#include <string>
#include <utility>
#include <vector>

using namespace std;

class Robot {
  static constexpr array<string, 4> directions = {"North", "East", "South",
                                                  "West"};
  vector<array<int, 3>> pos_dire;
  int i = 0;
  bool moved = false;

public:
  Robot(int width, int height) {
    for (int i = 0; i < width - 1; i++) {
      array<int, 3> arr = {i, 0, 1};
      if (i == 0) {
        arr[2] = 2;
      }
      pos_dire.push_back(arr);
    }
    for (int i = 0; i < height - 1; i++) {
      array<int, 3> arr = {width - 1, i, 0};
      if (i == 0) {
        arr[2] = 1;
      }
      pos_dire.push_back(arr);
    }
    for (int i = 0; i < width - 1; i++) {
      array<int, 3> arr = {width - 1 - i, height - 1, 3};
      if (i == 0) {
        arr[2] = 0;
      }
      pos_dire.push_back(arr);
    }
    for (int i = 0; i < height - 1; i++) {
      array<int, 3> arr = {0, height - 1 - i, 2};
      if (i == 0) {
        arr[2] = 3;
      }
      pos_dire.push_back(arr);
    }
  }

  void step(int num) {
    moved = true;
    int n = pos_dire.size();
    i = (i + num) % n;
  }

  vector<int> getPos() { return vector<int>{pos_dire[i][0], pos_dire[i][1]}; }

  string getDir() {
    if (moved) {
      return directions.at(pos_dire[i][2]);
    } else {
      return "East";
    }
  }
};

int main() {

  Robot *obj = new Robot(6, 3);
  obj->step(2);
  obj->step(2);
  obj->step(2);
  obj->step(1);
  obj->step(4);
  vector<int> param_2 = obj->getPos();
  string param_3 = obj->getDir();

  return 0;
}

File: leetcode_ts/2069.ts
class Robot2 {
  ss: string[];
  pos: number[];
  width: number;
  height: number;
  direction: number;
  constructor(width: number, height: number) {
    this.ss = ["North", "East", "South", "West"];
    this.width = width;
    this.height = height;
    this.direction = 1;
    this.pos = [0, 0];
  }

  check(): boolean {
    if (this.pos[0] >= this.width || this.pos[0] < 0) {
      return false;
    }
    if (this.pos[1] >= this.height || this.pos[1] < 0) {
      return false;
    }
    return true;
  }

  step(num: number): void {
    const funcs: (() => void)[] = [
      () => this.pos[1]++,
      () => this.pos[0]++,
      () => this.pos[1]--,
      () => this.pos[0]--,
    ];
    while (num > 0) {
      funcs[this.direction]();
      while (!this.check()) {
        funcs[(this.direction + 2) % 4]();
        this.direction = (this.direction + 3) % 4;
        funcs[this.direction]();
      }
      num -= 1;
    }
  }

  getPos(): number[] {
    return this.pos;
  }

  getDir(): string {
    return this.ss[this.direction];
  }
}

class Robot {
  private moved: boolean = false;
  private idx: number = 0;
  private pos: number[][] = [];
  private dir: number[] = [];
  private toDir: Map<number, string> = new Map();

  constructor(width: number, height: number) {
    this.toDir.set(0, "East");
    this.toDir.set(1, "North");
    this.toDir.set(2, "West");
    this.toDir.set(3, "South");

    for (let i = 0; i < width; ++i) {
      this.pos.push([i, 0]);
      this.dir.push(0);
    }
    for (let i = 1; i < height; ++i) {
      this.pos.push([width - 1, i]);
      this.dir.push(1);
    }
    for (let i = width - 2; i >= 0; --i) {
      this.pos.push([i, height - 1]);
      this.dir.push(2);
    }
    for (let i = height - 2; i > 0; --i) {
      this.pos.push([0, i]);
      this.dir.push(3);
    }
    this.dir[0] = 3;
  }

  step(num: number): void {
    this.moved = true;
    this.idx = (this.idx + num) % this.pos.length;
  }

  getPos(): number[] {
    return this.pos[this.idx];
  }

  getDir(): string {
    if (!this.moved) {
      return "East";
    }
    return this.toDir.get(this.dir[this.idx])!;
  }
}

const r = new Robot2(6, 3);
r.step(2);
r.step(2);
r.step(2);
r.step(1);
r.step(4);
console.log(r.getPos());
console.log(r.getDir());
